airflow_dags/open_elections/load_by_state.py

```diff
@@ -0,0 +1,402 @@
+        return enriched_df
+
+    @classmethod
+    def _clean_df(cls, df: pd.DataFrame) -> pd.DataFrame:
+        # Takes care of a number of issues:
+        #   - spaces appearing at the end of column names
+        #   - remove columns named "Unnamed : #" that come from empty CSV headers (normally from including index row)
+        #   - downcase Reg_voters column name
+        temp = df.copy()
+
+        for column in temp.columns:
+            if column.startswith('Unnamed'):
+                temp = temp.drop([column], axis=1)
+
+        clean_cols = {col: col.rstrip().lower() for col in temp.columns}
+        clean_cols_df = temp.rename(columns=clean_cols)
+
+        # Clean up the district column so it can be used for joins
+        def mapper(value: Any):
+            if pd.isna(value):
+                return DEFAULT_PK_VALUE
+            elif type(value) in (float, int):
+                return str(int(value))
+            elif type(value) == str:
+                return value
+            else:
+                raise ValueError('The value {} is not valid for district column'.format(value))
+
+        district_cleaned = clean_cols_df.assign(district=clean_cols_df['district'].map(mapper),
+                                                office=clean_cols_df['office'].map(mapper))
+
+        return district_cleaned
+
+
+class PrecinctFile(VoteFile):
+    SPLIT_NAME_LENGTH = 5
+
+
+class CountyFile(VoteFile):
+    SPLIT_NAME_LENGTH = 3
+
+
+class OfficeFile(VoteFile):
+    SPLIT_NAME_LENGTH = 4
+
+
+class StateMetadata:
+
+    def __init__(self,
+                 source_dir: str,
+                 state: str,
+                 vote_count_cols: List[str],
+                 county_pks: List[str],
+                 precinct_pks: List[str]):
+        self.source_dir = source_dir
+        self.state = state
+        self.vote_count_cols = vote_count_cols
+        self.county_pks = county_pks
+        self.precinct_pks = precinct_pks
+
+
+def load_to_dolt(dolt_dir: str,
+                 state_metadata: StateMetadata,
+                 load_elections: bool,
+                 load_precinct_votes: bool,
+                 load_county_votes: bool):
+    all_elections, all_precinct_votes, all_county_votes = parse_for_state(state_metadata)
+
+    repo = Dolt(dolt_dir)
+
+    if load_elections:
+        logger.info('Loading election data to "elections" in Dolt repo in {}'.format(dolt_dir))
+        import_list(repo, 'elections', all_elections, import_mode='update')
+    if load_precinct_votes:
+        logger.info('Loading election data to "{}_precinct_votes" in Dolt repo in {}'.format(state_metadata.state,
+                                                                                             dolt_dir))
+        import_list(repo,
+                    '{}_precinct_votes'.format(state_metadata.state),
+                    all_precinct_votes,
+                    import_mode='update',
```

Comment:
You sure you want update and not replace? What if the upstream deletes a bad row?

Reply:
I have not gotten that far yet, just parsing the data consistently proved to be a challenge.